sapicore/pipeline/simulation.py

```diff
@@ -0,0 +1,172 @@
+""" Simulate a network from a configuration YAML. """
+import os
+import logging
+
+from argparse import ArgumentParser
+from datetime import datetime
+
+import torch
+from torch import Tensor
+from alive_progress import alive_bar
+
+from sapicore.pipeline import Pipeline
+from sapicore.model import Model
+from sapicore.engine.network import Network
+
+from sapicore.utils.seed import fix_random_seed
+from sapicore.utils.signals import extend_input_current
+from sapicore.utils.tensorboard import TensorboardWriter, HDFData
+from sapicore.utils.constants import DT, TIME_FORMAT, SEED
+from sapicore.utils.io import ensure_dir, save_yaml, log_settings
+
+from sapicore.tests import ROOT
+
+# default hardware device for this pipeline.
+DEVICE = "cpu" if not torch.cuda.is_available() else "cuda:0"
+
+
+class Simulator(Pipeline):
+    """A prototypical simulation pipeline, bringing together model specification, disk I/O, and visualization.
+    Initialize a network and configure the simulation by providing a dictionary or a path to a YAML.
+
+    Parameters
+    ----------
+    configuration: dict or str
+        Configuration dictionary or path to a configuration YAML. If the path provided is relative, the simulation
+        pipeline will assume that the root is `sapicore/tests` to support the execution of YAML-configured tests.
+
+    """
+
+    def __init__(self, configuration: dict | str = None, **kwargs):
+        super().__init__(configuration=configuration, **kwargs)
+
+        # set random seed for entire simulation and apply default logger configuration.
+        fix_random_seed(SEED)
```

Comment:
This should probably be in its own method, e.g. `init_sim` that is called from init so that it can be overwritten by subclasses. But it's not super important.

Reply:
I'm slightly also weary that SEED is a constant. Because it would be nice if it's a constant only by default but can be overwritten by passing a different value to init or as part of the config.

Reply:
+1 for both. Wonder if `seed` and `device` are general enough to merit their own command line arguments in simulation.py. Regardless, I'll look these up in the config dict and override the defaults if they exist.

Reply:
`seed` at least should be a sim arg so we can iterate on it easily to estimate model variance with seed for example.

Reply:
Fixed. Leaving this in the constructor for now, but I'm open to adding a dedicated `initialize` method later:

```python
# set random seed for entire simulation based on config, kwarg, or default (in this order of preference).
random_seed = kwargs["seed"] if "seed" in kwargs.keys() else self.configuration.get("seed", SEED)
fix_random_seed(random_seed)
```